src/platform/w3dengine/client/w3dterrainvisual.cpp

```diff
@@ -0,0 +1,628 @@
+    }
+}
+
+void W3DTerrainVisual::Get_Water_Transform(const WaterHandle *handle, Matrix3D *transform)
+{
+    if (m_waterRenderObj != nullptr) {
+        m_waterRenderObj->Get_Grid_Transform(transform);
+    }
+}
+
+void W3DTerrainVisual::Set_Water_Grid_Resolution(
+    const WaterHandle *handle, float grid_cells_x, float grid_cells_y, float cell_size)
+{
+    if (m_waterRenderObj != nullptr) {
+        m_waterRenderObj->Set_Grid_Resolution(grid_cells_x, grid_cells_y, cell_size);
+    }
+}
+
+void W3DTerrainVisual::Get_Water_Grid_Resolution(
+    const WaterHandle *handle, float *grid_cells_x, float *grid_cells_y, float *cell_size)
+{
+    if (m_waterRenderObj != nullptr) {
+        m_waterRenderObj->Get_Grid_Resolution(grid_cells_x, grid_cells_y, cell_size);
+    }
+}
+
+void W3DTerrainVisual::Change_Water_Height(float wx, float wy, float delta)
+{
+    if (m_waterRenderObj != nullptr) {
+        m_waterRenderObj->Change_Grid_Height(wx, wy, delta);
+    }
+}
+
+void W3DTerrainVisual::Add_Water_Velocity(float world_x, float world_y, float z_velocity, float preferred_height)
+{
+    if (m_waterRenderObj != nullptr) {
+        m_waterRenderObj->Add_Velocity(world_x, world_y, z_velocity, preferred_height);
+    }
+}
+
+bool W3DTerrainVisual::Get_Water_Grid_Height(float x, float y, float *height)
+{
+    if (!m_waterGridEnabled) {
+        return false;
+    }
+
+    if (m_waterRenderObj == nullptr) {
+        return false;
+    }
+
+    float gx;
+    float gy;
+
+    if (!m_waterRenderObj->World_To_Grid_Space(x, y, gx, gy)) {
+        return false;
+    }
+
+    m_waterRenderObj->Get_Grid_Vertex_Height(GameMath::Fast_To_Int_Truncate(gx), GameMath::Fast_To_Int_Truncate(gy), height);
+    return true;
+}
+
+void W3DTerrainVisual::Set_Terrain_Tracks_Detail()
+{
+    if (g_theTerrainTracksRenderObjClassSystem != nullptr) {
+        g_theTerrainTracksRenderObjClassSystem->Set_Detail();
+    }
+}
+
+void W3DTerrainVisual::Set_Shore_Line_Detail()
+{
+    if (m_baseHeightMap != nullptr) {
+        m_baseHeightMap->Set_Shore_Line_Detail();
+    }
+}
+
+void W3DTerrainVisual::Add_Faction_Bib(Object *object, bool highlighted, float width)
+{
+    if (m_heightMap != nullptr) {
+        float exit_width = object->Get_Template()->Get_Factory_Exit_Width();
+        float bib_width = width + object->Get_Template()->Get_Factory_Extra_Bib_Width();
+        GeometryInfo info = object->Get_Geometry_Info();
+        float major_radius = info.Get_Major_Radius();
+        float minor_radius = info.Get_Minor_Radius();
+
+        if (info.Get_Type() != GEOMETRY_BOX) {
+            minor_radius = major_radius;
+        }
+
+        Vector3 points[4];
+        points[0].X = bib_width + major_radius;
+        points[0].Y = minor_radius + bib_width;
+        points[0].Z = 0.0f;
+
+        points[1].X = major_radius + exit_width + bib_width;
+        points[1].Y = minor_radius + bib_width;
+        points[1].Z = 0.0f;
+
+        points[2].X = points[1].X;
```

Comment:
Right now it looks like points 0,3 and 1,2 share same positions. Is this correct?

Same in `W3DTerrainVisual::Add_Faction_Bib_Drawable` function.

Reply:
Fixed.